diff --git a/src/http/modules/ngx_http_ssl_module.c b/src/http/modules/ngx_http_ssl_module.c
index df1e55e..497a49e 100644
--- a/src/http/modules/ngx_http_ssl_module.c
+++ b/src/http/modules/ngx_http_ssl_module.c
@@ -293,7 +293,7 @@ static ngx_http_variable_t  ngx_http_ssl_vars[] = {
 };
 
 
-static ngx_str_t ngx_http_ssl_sess_id_ctx = ngx_string("HTTP");
+ngx_str_t ngx_http_ssl_sess_id_ctx = ngx_string("HTTP");
 
 
 #ifdef TLSEXT_TYPE_application_layer_protocol_negotiation
diff --git a/src/http/modules/ngx_http_ssl_module.h b/src/http/modules/ngx_http_ssl_module.h
index ec2c62f..894df7d 100644
--- a/src/http/modules/ngx_http_ssl_module.h
+++ b/src/http/modules/ngx_http_ssl_module.h
@@ -13,6 +13,7 @@
 #include <ngx_core.h>
 #include <ngx_http.h>
 
+extern ngx_str_t ngx_http_ssl_sess_id_ctx;
 
 typedef struct {
     ngx_flag_t                      enable;
diff --git a/src/http/ngx_http_request.c b/src/http/ngx_http_request.c
index 0bb1b8c..02e18f9 100644
--- a/src/http/ngx_http_request.c
+++ b/src/http/ngx_http_request.c
@@ -63,6 +63,9 @@ static void ngx_http_ssl_handshake(ngx_event_t *rev);
 static void ngx_http_ssl_handshake_handler(ngx_connection_t *c);
 #endif
 
+#if (NGX_HTTP_VIRTUALSSL)
+#include <ngx_http_virtualssl_module.h>
+#endif
 
 static char *ngx_http_client_errors[] = {
 
@@ -765,6 +768,10 @@ ngx_http_ssl_servername(ngx_ssl_conn_t *ssl_conn, int *ad, void *arg)
     ngx_http_core_loc_conf_t  *clcf;
     ngx_http_core_srv_conf_t  *cscf;
 
+#if (NGX_HTTP_VIRTUALSSL)
+    int                       ret;
+#endif
+
     servername = SSL_get_servername(ssl_conn, TLSEXT_NAMETYPE_host_name);
 
     if (servername == NULL) {
@@ -788,6 +795,12 @@ ngx_http_ssl_servername(ngx_ssl_conn_t *ssl_conn, int *ad, void *arg)
         return SSL_TLSEXT_ERR_NOACK;
     }
 
+#if (NGX_HTTP_VIRTUALSSL)
+    ret = ngx_http_virtualssl_handle_sni(c, &host);
+     if (ret != SSL_TLSEXT_ERR_NOACK)
+         return ret;
+#endif
+
     hc = c->data;
 
     if (ngx_http_find_virtual_server(c, hc->addr_conf->virtual_names, &host,
